core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala

```diff
@@ -234,8 +244,22 @@ private[spark] class EventLoggingListener(
     }
   }
 
-  // No-op because logging every update would be overkill
-  override def onExecutorMetricsUpdate(event: SparkListenerExecutorMetricsUpdate): Unit = { }
+  /**
+   * Log if there is a new peak value for one of the memory metrics for the given executor.
+   * Metrics are cleared out when a new stage is started in onStageSubmitted, so this will
+   * log new peak memory metric values per executor per stage.
+   */
+  override def onExecutorMetricsUpdate(event: SparkListenerExecutorMetricsUpdate): Unit = {
```

Comment:
For a longer running stage, once it ramps up, hopefully there wouldn't be a lot of new peak values. Looking at a subset of our applications, the extra logging overhead has mostly been between 0.25% to 1%, but it can be 8%. 

By logging each peak value at the time they occur (and reinitializing when a stage starts), it's possible to tell which stages are active at the time, and it would potentially be possible to graph these changes on a timeline -- this information wouldn't be available if the metrics are only logged at stage end, and the times are lost.

Logging at stage end would limit the amount of extra logging. If we add more metrics (such as for offheap), then there could be more new peaks and more extra logging with the current approach. Excess logging is a concern, and I can move to stage end if the overhead is too much.

Reply:
You could tell which stages are active from the other events.  Do you think that timeline would be useful?  Since its only keeping the peak, I'm not sure how interesting that graph is.

With overlapping stages, you might lose some information by only logging at stage end -- eg. if first stage 1 is running for a while, with peak X, and then stage 2 starts up with peak Y > X, you would only ever log peak Y for stage 1.  You could address this by also logging at stage start, but then you're back to more logging (and its really tricky to figure out a minimal set of things of log when stage 2 starts, as you don't know which executor its going to run on).

But I'm still not sure what we'd do with those extra values, and if there is any value in capturing them.

(some this design discussion probably belongs on jira -- I'll also go reread the design doc now)

Reply:
Also on log size -- was there anything special about the 8% case?  Eg. was a tiny application running on a ton of executors, so the logs were small to begin with?  If so, then its probably fine.

Reply:
We're not constructing a timeline currently, and yes, with only peak values, it could be rather sparse. We would get new values with new stages, but would not see decreases in memory during a stage.

The situation where there is a stage 1 with peak X, and then stage 2 starts with peak Y > X is interesting though, and it would be useful to have this information, since we would then know to focus on stage 2 for memory consumption, even though both 1 and 2 could have the same peak values. Logging at stage start would double the amount of logging, and would be trickier, so I'd prefer either keeping the current approach or only logging at stage end.

The higher logging was for smaller applications (and smaller logs).

Reply:
ok that makes sense to me.  If the logging overhead is small, logging on every new peak certainly seems simpler.

Reply:
yeah logging an event per executor at stage end seems good to me.  It would be great if we could see how much that version affects log size as well, if you can get those metrics.

also these tradeoffs should go into the design doc, its harder to find comments from a PR after this feature has been merged.  For now, it would also be nice if you could post a version that everyone can comment on, eg. a google doc.

Reply:
ExecutorMetrics right now has: jvmUsedHeapMemory, jvmUsedNonHeapMemory, onHeapExecutionMemory, offHeapExecutionMemory, onHeapStorageMemory, and offHeapStorageMemory. For logging at stage end, we can log the peak for each of these, but unified memory is more problematic. We could add new fields for on heap/off heap unified memory, but I'm inclined to remove unified memory (from all the places it is currently used), rather than add more fields. Users can still sum peak execution and peak storage values, which may be larger than the actual peak unified memory if they are not at peak values at the same time, but should still be a reasonable estimate for sizing.

Reply:
is your concern the size of the msg from the executors to the driver?

that certainly is valid, but I wonder if we should think a bit harder about this if that is going to be a common concern, as I think we'll want to add more metrics.

One possibility is for the executor to do the peak calculation itself, and then only send an update for the metrics with a new peak.  Also that would let us just send the peak on task end events.

I'm just brainstorming at the moment, not saying it should be changed one way or the other ... need to think about it more

Reply:
Thanks for your thoughts on this. Size of message, and also logging, but it is only an extra few longs per heartbeat, and and similarly for logging. Task end would help with minimizing communication for longer running tasks. The heartbeats are only every 10 seconds, so perhaps not so bad.